EpikCord/client/http.py

```diff
@@ -0,0 +1,259 @@
+        self.global_event.set()
+
+    async def ws_connect(
+        self, url: str, **kwargs
+    ) -> aiohttp.ClientWebSocketResponse:
+        return await self.session.ws_connect(url, **kwargs)
+
+    async def request(
+        self,
+        route: Route,
+        *args,
+        discord: bool = True,
+        json: Optional[Dict[str, Any]] = None,
+        files: Optional[List[File]] = None,
+        **kwargs,
+    ) -> aiohttp.ClientResponse:
+        """
+        Parameters
+        ----------
+        *args
+            The args to pass to the ClientSession.request method.
+        discord: bool
+            Whether this request is to Discord.
+        json: Optional[Dict[:class:`str`, Any]]
+            The json document to pass to the ClientSession.request method.
+        files: Optional[List[File]]
+            The files to pass to the request.
+        **kwargs
+            The kwargs to pass to the ClientSession.request method.
+
+        Returns
+        -------
+        aiohttp.ClientResponse
+            The response of the request.
+
+        Raises
+        ------
+        NotFound
+            If the request returns a 404 status code.
+        Forbidden
+            If the request returns a 403 status code.
+        Unauthorized
+            If the request returns a 401 status code.
+        BadRequest
+            If the request returns a 400 status code.
+        HTTPException
+            If the request returns a status code that is not OK and not any of the other exceptions.
+        """
+
+        method = route.method
+        url = route.url
+
+        if not discord:
+            return await self.session.request(method, url, *args, **kwargs)
+
+        self.setup_kwargs(kwargs, files=files, json=json)
+
+        url = clean_url(url, self.version)
+        bucket = self.buckets.get(f"{method}:{url}") or MockBucket()
+
+        await self.global_event.wait()
+        await bucket.wait()
+
+        for _ in range(5):
+            async with self.session.request(method, url, *args, **kwargs) as response:
+                if isinstance(bucket, MockBucket):
+                    bucket = await self.set_bucket(
+                        response=response,
+                        **route.major_parameters,
+                    )
+
+                data = await extract_content(response)
+
+                await self.log_request(response, data)
+
+                if response.status == 429:
+                    await self.handle_ratelimit(data, bucket)
+                elif response.headers.get("X-RateLimit-Remaining", "1") == "0":
+                    await bucket.handle_exhaustion(int(response.headers["X-RateLimit-Reset-After"]))
+                elif not response.ok:
+                    error = self.error_mapping.get(response.status, HTTPException)
+                    raise error(response, data)
+                elif response.ok:
+                    return response
```

Comment:
The response part should be extracted outside `request`

Reply:
Why? It looks fine.